### PR TITLE
expression: fix issue that `TimeIsNull` is not compatible with MySQL

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/9763.

### What is changed and how it works?
Introduce a new variable `isNotNull` to `builtinTimeIsNullSig` to judge if this column has a `NotNull` attribution and push it down to `TiKV`.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/distsql_builtin.go [hunk 1]
```diff
@@ -336,7 +336,12 @@ func getSignatureByPB(ctx sessionctx.Context, sigCode tipb.ScalarFuncSig, tp *ti
 	case tipb.ScalarFuncSig_RealIsNull:
 		f = &builtinRealIsNullSig{base}
 	case tipb.ScalarFuncSig_TimeIsNull:
-		f = &builtinTimeIsNullSig{base}
+		isNotNull := false
+		if len(expr.Children) > 0 && mysql.HasNotNullFlag(uint(expr.Children[0].FieldType.Flag)) {
```
Comment (breeswish) [inline expression/distsql_builtin.go]: No need for this check any more?

File: expression/builtin_op.go [hunk 2]
```diff
@@ -799,7 +799,11 @@ func (c *isNullFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 		sig = &builtinRealIsNullSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_RealIsNull)
 	case types.ETDatetime:
-		sig = &builtinTimeIsNullSig{bf}
+		isNotNull := false
+		if mysql.HasNotNullFlag(args[0].GetType().Flag) {
```
Comment (lamxTyler) [inline expression/builtin_op.go]: Is it possible that the `NotNull` flag is generated from expressions, for example `date(a)`? Should we also specially handle it?

Comment (qw4990) [reply expression/builtin_op.go]: How about add a more condition that only when `args[0]` is `Column` we can set `isNotNull` to ture?

Comment (breeswish) [reply expression/builtin_op.go]:
```
mysql> select c from t;
Field   1:  `c`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       LONG
Collation:  binary (63)
Length:     11
Max_length: 1
Decimals:   0
Flags:      NOT_NULL NO_DEFAULT_VALUE NUM 


+---+
| c |
+---+
| 0 |
+---+
1 row in set (0.00 sec)

mysql> select date(c) from t;
Field   1:  `date(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATE
Collation:  binary (63)
Length:     10
Max_length: 10
Decimals:   0
Flags:      BINARY 


+------------+
| date(c)    |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)
```

Looks like expression clears the not null flag?

Comment (lamxTyler) [reply expression/builtin_op.go]: I think it is ok.

Comment (qw4990) [reply expression/builtin_op.go]: @breeswish 
Try this case.
```
create table t (a datetime not null);
insert ignore into t values (0);
select * from t where a is null;
select * from t where date(a) is null;
```

In MySQL, the latter query returns nothing.

Comment (breeswish) [reply expression/builtin_op.go]: Yes. I mean, it looks like `a` is NOT NULL does not mean that `date(a)` is NOT NULL, as indicated in https://github.com/pingcap/tidb/pull/10957#discussion_r298932689 . So this does not mean that this special rule applies only to columns.

Comment (breeswish) [reply expression/builtin_op.go]: Actually looks like there is another TiDB bug..

```sql
create table t (a datetime not null);
insert ignore into t values (0);
select date(a) from t;
```

TiDB:

```
+---------+
| date(a) |
+---------+
| NULL    |
+---------+
1 row in set, 1 warning (0.01 sec)
```

MySQL:

```
+------------+
| date(a)    |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)
```

Also the result of TiDB also indicates that `date(a)` can be NULL even if `a` is NOT NULL..

Comment (qw4990) [reply expression/builtin_op.go]: There are lots of bugs(or differences) between TiDB and MySQL caused by different implementation details about type system and scalar functions...

Comment (qw4990) [reply expression/builtin_op.go]: There may be another scalar function `f(a)` which can propagate the `NOT NULL` attribution from `a` to its outer world.
Check expression type and make it works only when it is a `Column` is safer.

File: expression/distsql_builtin.go (hunk 1 above)
Comment (qw4990) [reply expression/distsql_builtin.go]: Which check? `len(expr.Children) > 0`?.

File: expression/builtin_op.go (hunk 2 above)
Comment (breeswish) [reply expression/builtin_op.go]: This column check may not work in TiKV..

Comment (breeswish) [reply expression/builtin_op.go]: You will have no way to know the underlying expression is a column or not.

Comment (qw4990) [reply expression/builtin_op.go]: Is this comment can be resolved?
I will fix the check problem in TiKV later.

**Comment by qw4990** [comment]
@lamxTyler @breeswish PTAL~

File: expression/builtin_op.go [hunk 3]
```diff
@@ -799,7 +799,12 @@ func (c *isNullFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 		sig = &builtinRealIsNullSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_RealIsNull)
 	case types.ETDatetime:
-		sig = &builtinTimeIsNullSig{bf}
+		isNotNull := false
+		_, isCol := args[0].(*Column)
```
Comment (lonng) [inline expression/builtin_op.go]:
```suggestion
		col, isCol := args[0].(*Column)
```

File: expression/builtin_op.go [hunk 4]
```diff
@@ -799,7 +799,12 @@ func (c *isNullFunctionClass) getFunction(ctx sessionctx.Context, args []Express
...
+		if isCol && mysql.HasNotNullFlag(args[0].GetType().Flag) {
```
Comment (lonng) [inline expression/builtin_op.go]:
```suggestion
		if isCol && mysql.HasNotNullFlag(col.GetType().Flag) {
```
There is no need to index the slice again.

**Comment by lonng** [comment]
You should override the `implicitArgs` for this function and push it down to TiKV according to your description. But I cannot find the code snippet corresponding.

**Comment by qw4990** [comment]
The way in this PR to fix this issue is not appropriate, so I close it and fix it by another PR.

**Pull request closed without merging**